File: libs/eave-stdlib-py/src/eave/stdlib/core_api/operations/forge.py
from typing import Optional
import uuid
from eave.stdlib.core_api.models import team

from eave.stdlib.core_api.models.forge import ForgeInstallation, QueryForgeInstallationInput
from eave.stdlib.core_api.models.forge import RegisterForgeInstallationInput
from eave.stdlib.core_api.models.forge import UpdateForgeInstallationInput
from eave.stdlib.core_api.operations import BaseRequestBody, BaseResponseBody, Endpoint, EndpointConfiguration
from . import BaseRequestBody, BaseResponseBody, Endpoint, EndpointConfiguration

from eave.stdlib.eave_origins import EaveOrigin

from ..models import team
from ... import requests


class RegisterForgeInstallation(Endpoint):
    config = EndpointConfiguration(
        path="/integrations/forge/register",
        auth_required=False,
        team_id_required=False,
        signature_required=False,  # FIXME: Get signing working from Forge
        origin_required=True,
    )

    class RequestBody(BaseRequestBody):
        forge_integration: RegisterForgeInstallationInput

    class ResponseBody(BaseResponseBody):
        forge_integration: ForgeInstallation

    @classmethod
    async def perform(cls,
        origin: EaveOrigin,
        input: RequestBody,
    ) -> ResponseBody:
        response = await requests.make_request(
            url=cls.config.url,
            origin=origin,
            input=input,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json)


class UpdateForgeInstallation(Endpoint):
    config = EndpointConfiguration(
        path="/integrations/forge/update",
        auth_required=False,
        team_id_required=False,
        signature_required=True,
        origin_required=True,
    )

    class RequestBody(BaseRequestBody):
        forge_integration: UpdateForgeInstallationInput

    class ResponseBody(BaseResponseBody):
        team: Optional[team.Team]
        forge_integration: ForgeInstallation


    @classmethod
    async def perform(cls,
        origin: EaveOrigin,
        input: RequestBody,
    ) -> ResponseBody:
        response = await requests.make_request(
            url=cls.config.url,
            origin=origin,
            input=input,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json)


    @classmethod
    async def perform_authed(cls,
        origin: EaveOrigin,
        account_id: uuid.UUID,
        access_token: str,
        input: RequestBody,
    ) -> ResponseBody:
        response = await requests.make_request(
            url=cls.config.url,
            origin=origin,
            input=input,
            access_token=access_token,
            account_id=account_id,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json)






class QueryForgeInstallation(Endpoint):
    config = EndpointConfiguration(
        path="/integrations/forge/query",
        auth_required=False,
        team_id_required=False,
        signature_required=True,
        origin_required=True,
    )

    class RequestBody(BaseRequestBody):
        forge_integration: QueryForgeInstallationInput

    class ResponseBody(BaseResponseBody):
        team: Optional[team.Team]
        forge_integration: ForgeInstallation

    @classmethod
    async def perform(cls,
        origin: EaveOrigin,
        input: RequestBody,
    ) -> ResponseBody:
        response = await requests.make_request(
            url=cls.config.url,
            origin=origin,
            input=input,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json)

File: libs/eave-stdlib-ts/src/core-api/operations/forge.ts
import { RequestArgsOrigin, makeRequest } from '../../lib/requests.js';
import { sharedConfig } from '../../config.js';
import { ForgeInstallation, QueryForgeInstallationInput, RegisterForgeInstallationInput, UpdateForgeInstallationInput } from '../models/forge.js';
import { Team } from '../models/team.js';


export type QueryForgeInstallationRequestBody = {
  forge_integration: QueryForgeInstallationInput;
}

export type QueryForgeInstallationResponseBody = {
  team: Team;
  forge_integration: ForgeInstallation;
}
export type RegisterForgeInstallationRequestBody = {
  forge_integration: RegisterForgeInstallationInput;
}
export type RegisterForgeInstallationResponseBody = {
  forge_integration: ForgeInstallation;
}

export type UpdateForgeInstallationRequestBody = {
  forge_integration: UpdateForgeInstallationInput;
}
export type UpdateForgeInstallationResponseBody = {
  team: Team;
  forge_integration: ForgeInstallation;
}

export async function queryForgeInstallation({ origin, input }: RequestArgsOrigin & {input: QueryForgeInstallationInput}): Promise<QueryForgeInstallationResponseBody> {
  const resp = await makeRequest({
    origin,
    url: `${sharedConfig.eaveApiBase}/integrations/forge/query`,
    input,
  });
  const responseData = <QueryForgeInstallationResponseBody>(await resp.json());
  return responseData;
}

export async function registerForgeInstallationInsecure({ origin, sharedSecret, input }: RequestArgsOrigin & {sharedSecret: string, input: RegisterForgeInstallationRequestBody}): Promise<RegisterForgeInstallationResponseBody> {
  const resp = await makeRequest({
    origin,
    accessToken: sharedSecret,
    sign: false, // TODO: Get signing working
    url: `${sharedConfig.eaveApiBase}/integrations/forge/register`,
    input,
  });
  const responseData = <RegisterForgeInstallationResponseBody>(await resp.json());
  return responseData;
}

export async function updateForgeInstallationInsecure({ origin, sharedSecret, input }: RequestArgsOrigin & {sharedSecret: string, input: UpdateForgeInstallationRequestBody}): Promise<UpdateForgeInstallationResponseBody> {
  const resp = await makeRequest({
    origin,
    accessToken: sharedSecret,
    sign: false,
    url: `${sharedConfig.eaveApiBase}/integrations/forge/update`,
    input,
  });
  const responseData = <UpdateForgeInstallationResponseBody>(await resp.json());
  return responseData;
}
